File: apps/mo2-android/android/core/common/src/main/kotlin/jp/developershub/dub/mo2/core/common/SessionStore.kt
// SessionStore — client-side secret vault contract (mirror of session-store.ts).
// Holds a single opaque session token (theme8, no 2-token/JWT), the mobile
// refresh token, and the server-issued deviceId ULID. Never logged. The Android
// impl is EncryptedDataStore (core:database); InMemorySessionStore backs tests.
package jp.developershub.dub.mo2.core.common

interface SessionStore {
    fun getToken(): String?
    fun getRefreshToken(): String?
    fun getDeviceId(): String?
    fun setSession(token: String, refreshToken: String?)
    fun setDeviceId(deviceId: String)
    fun clear() // logout
}

/** In-memory impl for unit tests (and previews). Mirrors InMemorySessionStore. */
class InMemorySessionStore : SessionStore {
    private var token: String? = null
    private var refreshToken: String? = null
    private var deviceId: String? = null

    override fun getToken(): String? = token
    override fun getRefreshToken(): String? = refreshToken
    override fun getDeviceId(): String? = deviceId

    override fun setSession(token: String, refreshToken: String?) {
        this.token = token
        this.refreshToken = refreshToken
    }

    override fun setDeviceId(deviceId: String) {
        this.deviceId = deviceId
    }

    override fun clear() {
        token = null
        refreshToken = null
        // deviceId is kept — it survives logout (server ULID; §3).
    }
}

File: apps/mo2-android/src/session-store.ts
// SessionStore — the client-side secret vault abstraction. On Android the impl
// is EncryptedDataStore (§3); here we model the contract + an in-memory impl for
// tests. Holds a single opaque session token (theme8, no 2-token/JWT), the
// mobile refresh token, and the server-issued deviceId ULID. Never logged.
export interface SessionStore {
  getToken(): string | null;
  getRefreshToken(): string | null;
  getDeviceId(): string | null;
  setSession(token: string, refreshToken: string | null): void;
  setDeviceId(deviceId: string): void;
  clear(): void; // logout
}

export class InMemorySessionStore implements SessionStore {
  private token: string | null = null;
  private refreshToken: string | null = null;
  private deviceId: string | null = null;

  getToken(): string | null {
    return this.token;
  }
  getRefreshToken(): string | null {
    return this.refreshToken;
  }
  getDeviceId(): string | null {
    return this.deviceId;
  }
  setSession(token: string, refreshToken: string | null): void {
    this.token = token;
    this.refreshToken = refreshToken;
  }
  setDeviceId(deviceId: string): void {
    this.deviceId = deviceId;
  }
  clear(): void {
    this.token = null;
    this.refreshToken = null;
    // deviceId is kept — it survives logout (server ULID; §3).
  }
}
